### PR TITLE
chore(Storage): Simplify Storage.getURL

### PR DESCRIPTION
**Pull request by jcjimenez**
## Issue \#

https://github.com/aws-amplify/amplify-swift/issues/2727

## Description

This is a simplification of the StorageCategoryBehavior.getURL(options:) feature's internals. No public API changes were made.

## General Checklist

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [x] All integration tests pass
~~- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)~~
~~- [ ] Documentation update for the change if required~~
- [x] PR title conforms to conventional commit style
- [x] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
~~- [ ] If breaking change, documentation/changelog update with migration instructions~~

Screenshot 2023-02-01 at 8 22 22 AM

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Dependency/AWSS3PreSignedURLBuilderBehavior.swift [hunk 1]
```diff
@@ -23,15 +23,3 @@ protocol AWSS3PreSignedURLBuilderBehavior {
...
-
-    func getPreSignedURL(key: String, signingOperation: AWSS3SigningOperation, expires: Int64? = nil) async throws -> URL {
-        try await getPreSignedURL(key: key, signingOperation: signingOperation, expires: expires)
-    }
-
-    func getPreSignedURL(key: String, expires: Int64? = nil) async throws -> URL {
-        try await getPreSignedURL(key: key, signingOperation: .getObject, expires: expires)
-    }
-
-}
```
Comment (jcjimenez) [inline AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Dependency/AWSS3PreSignedURLBuilderBehavior.swift:37]: Removing these in the spirit of "explicit is better than implicit"

File: AmplifyPlugins/Storage/Tests/AWSS3StoragePluginTests/AWSS3StoragePluginAsyncBehaviorTests.swift [hunk 2]
```diff
@@ -47,21 +47,6 @@ class AWSS3StoragePluginAsyncBehaviorTests: XCTestCase {
         queue = nil
     }
 
-    func testPluginGetURLAsync() async throws {
```
Comment (jcjimenez) [inline AmplifyPlugins/Storage/Tests/AWSS3StoragePluginTests/AWSS3StoragePluginAsyncBehaviorTests.swift:50]: This test was moved to AWSS3StoragePluginGetPresignedUrlTests.

File: AmplifyPlugins/Storage/Tests/AWSS3StoragePluginTests/AWSS3StoragePluginGetPresignedUrlTests.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        storageService = nil
+        authService = nil
+        testKey = nil
+        testURL = nil
+        testData = nil
+        queue = nil
+        systemUnderTest = nil
+    }
+
+    func testPluginGetURLAsync() async throws {
```
Comment (jcjimenez) [inline AmplifyPlugins/Storage/Tests/AWSS3StoragePluginTests/AWSS3StoragePluginGetPresignedUrlTests.swift:57]: This came from AWSS3StoragePluginAsyncBehaviorTests

File: AmplifyPlugins/Storage/Tests/AWSS3StoragePluginTests/AWSS3StoragePluginGetPresignedUrlTests.swift [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        XCTAssertEqual(authService.interactions, [
+            "getIdentityID()"
+        ])
+        let expectedServiceKey = "public/" + testKey
+        XCTAssertEqual(storageService.interactions, [
+            "getPreSignedURL(serviceKey:signingOperation:expires:) \(expectedServiceKey) getObject 18000"
+        ])
+    }
+    
+    func testGetURLOperationValidationError() async throws {
```
Comment (jcjimenez) [inline AmplifyPlugins/Storage/Tests/AWSS3StoragePluginTests/AWSS3StoragePluginGetPresignedUrlTests.swift:70]: The rest of these are translations of the old NSOperation tests.

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Service/Storage/AWSS3StorageService+UploadBehavior.swift [hunk 5]
```diff
@@ -20,12 +20,13 @@ extension AWSS3StorageService {
...
         }
 
+        let bucket = self.bucket ?? ""
         guard attempt(try validateParameters(bucket: bucket, key: serviceKey, accelerationModeEnabled: false), fail: fail) else { return }
 
         Task {
-            let transferTask = createTransferTask(transferType: .upload(onEvent: onEvent),
-                                                  bucket: bucket,
-                                                  key: serviceKey)
+            let transferTask = self.createTransferTask(transferType: .upload(onEvent: onEvent),
```
Comment (atierian) [inline AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Service/Storage/AWSS3StorageService+UploadBehavior.swift:27]: Why add the explicit `self` here and not in the other places where self is implicitly accessed? I don't have an opinion on implicit or explicit here, but can we make it consistent?

Also, could there be any unintended consequences (e.g. Task outliving `self`) with the strong capture here?

Comment (jcjimenez) [reply AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Service/Storage/AWSS3StorageService+UploadBehavior.swift:27]: Thank you for catching this - I don't remember why the compiler was complaining, so I added these, but I just removed them and things are back to normal; Let me revert these.

Relating to the implicit or explicit retention of self within the `Task`, I think it's equivalent either way. As long as the task itself isn't retained by self, we should be fine. Or were you thinking about a possibility of unexpected behavior like when a unit test exits while a task/operation is still running?

Comment (atierian) [reply AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Service/Storage/AWSS3StorageService+UploadBehavior.swift]: Yea sorry I wasn't clear. Implicit vs explicit shouldn't make a difference with self retention; it was a general question (that is admittedly outside the scope of this change) about self being strongly captured there. But it doesn't appear to be a concern.

**Review by atierian — APPROVED** [review]

**Pull request merged**